src/platform/backends/qemu/firewall_config.cpp

```diff
@@ -220,54 +233,148 @@ void clear_iptables_rules_for(const QString& table, const QString& bridge_name,
             rule.remove(0, 3);
 
-            // Pass the chain and rule wholesale since we capture the whole line
-            delete_iptables_rule(table, QStringList() << rule);
+            try
+            {
+                // Pass the chain and rule wholesale since we capture the whole line
+                delete_firewall_rule(firewall, table, QStringList() << rule);
+            }
+            catch (const std::runtime_error& e)
+            {
+                mpl::log(mpl::Level::error, category,
+                         fmt::format("Error deleting firewall rule '{}': {}", rule, e.what()));
+            }
         }
     }
 }
-} // namespace
 
-mp::IPTablesConfig::IPTablesConfig(const QString& bridge_name, const std::string& subnet)
-    : bridge_name{bridge_name},
-      cidr{QString("%1.0/24").arg(QString::fromStdString(subnet))},
-      comment{multipass_iptables_comment(bridge_name)}
+auto is_firewall_in_use(const QString& firewall)
+{
+    QRegularExpression re{"^-[ARIN]"};
+
+    for (const auto& table : firewall_tables)
+    {
+        auto rules = get_firewall_rules(firewall, table);
+
+        for (const auto& line : rules.split('\n'))
+        {
+            if (re.match(line).hasMatch())
+            {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+// We require a >= 5.2 kernel to avoid weird conflicts with xtables and support for inet table NAT rules.
+// Taken from LXD :)
+void check_kernel_support()
 {
+    const auto kernel_version_parts{MP_UTILS.get_kernel_version().split('.')};
+
+    if (kernel_version_parts.size() < 2)
+    {
+        throw std::runtime_error("Failed converting kernel version into parts");
+    }
+
+    bool ok;
+    auto major_version = kernel_version_parts[0].toInt(&ok);
+    if (!ok)
+    {
+        throw std::runtime_error("Cannot parse kernel major number");
+    }
+
+    auto minor_version = kernel_version_parts[1].toInt(&ok);
+    if (!ok)
+    {
+        throw std::runtime_error("Cannot parse kernel minor number");
+    }
+
+    if (major_version < 5 || (major_version == 5 && minor_version < 2))
+    {
+        throw std::runtime_error("Kernel version does not meet minimum requirement of 5.2");
+    }
+}
+
+auto iptables_in_use()
+{
+    return is_firewall_in_use(iptables);
+}
+
+auto nftables_in_use()
+{
+    check_kernel_support();
+    return is_firewall_in_use(nftables);
+}
+
+auto detect_firewall()
+{
+    QString firewall_exec;
+
     try
     {
-        clear_all_iptables_rules();
-        set_iptables_rules(bridge_name, cidr, comment);
+        if (nftables_in_use())
+        {
+            firewall_exec = nftables;
+        }
+        else if (iptables_in_use())
+        {
+            firewall_exec = iptables;
+        }
+        else
+        {
+            firewall_exec = nftables;
+        }
     }
-    catch (const std::exception& e)
+    catch (const std::runtime_error& e)
```

Comment:
So, part of the problem here is that the `Process` class (and derived classes) throw `std::runtime_error` for all sorts of reasons, some of which are not fatal here.  As I don't think this is the proper place to start adding different exception types for a class otherwise not touched in this PR (and I've already had to touch other seemingly unrelated areas), I'm not sure how else to handle other than to catch `std::runtime_error` here.

For example, say `iptables-nft` does not exist and when the code executes that in the `Process` class, it will throw a `std::runtime_error`.  But it's not fatal that `iptables-nft` doesn't exist- we'll just default to use `iptables-legacy` which should exist.  However, if that doesn't exist, then it is fatal.  So other than catching `std::runtime_error` here or creating a whole new exception for this case in the `Process` class (which I already mentioned I'm reluctant to do), I'm not sure what else there is to do.  Any suggestions?

Reply:
I went back and it's not `Process` throwing the exception, it's in the firewall code.  So yeah, I'll add a new exception for process errors and catch that, but let other exceptions bubble up.  Sorry for the noise and thanks!

Reply:
Ok, done.